### PR TITLE
allow custom format on item for Debug

### LINKED ISSUE #277: Support custom format on item for `Debug` derive macro
I.e. something like:

```rs
#[derive(Debug)]
#[debug("{name}={value:?}")]
struct Arg {
  name: String,
  value: String
}
```

### PR DESCRIPTION
**Pull request by ModProg**
Resolves #277

## Checklist

- [x] Documentation is updated (if required)
- [x] Tests are added/updated (if required)
- [x] [CHANGELOG entry][l:1] is added (if required)

[l:1]: /CHANGELOG.md

### CODE REVIEW

**Comment by tyranron** [comment]
@ModProg @JelteF also, it's a bit unclear how it should work with the item-level formatting attributes?

```rust
#[derive(Debug)]
#[debug("{name}={value:?}")]
struct Arg {
  #[debug("{name}!")]
  name: String,
  value: String
}
```
What should happen here?
1. Compilation error saying that we cannot mix top-level and field-level formatting attributes.
2. Print `name!=value`, considering the field-level formatting in top-level attribute.
3. Print `name=value`, silently ignoring the field-level formatting.

I'd vote for the option 1.

**Comment by JelteF** [comment]
> it's a bit unclear how it should work with the item-level formatting attributes?

I think option 1 seems good.

**Comment by ModProg** [comment]
while 1. is definitly easier, 2. looks reasonable as well. but we can do that later, as it's a non breaking change.

File: tests/debug.rs [hunk 1]
```diff
@@ -1153,3 +1153,41 @@ mod generic {
...
+
+    #[test]
+    fn enums() {
+        #[derive(Debug)]
+        #[debug("Format String")]
+        enum Item {
+            Unit,
+        }
+        assert_eq!(format!("{:?}", Item::Unit), "Format String");
+    }
```
[the remark this answers is not in the document]
Comment (ModProg) [reply tests/debug.rs:1192]: Maybe it is even the other way round, format on the enum itself should probably match #142, meaning that this should only do variant formatting but not item formatting.

[the remark this answers is not in the document]
Comment (tyranron) [reply tests/debug.rs:1192]: @ModProg yes... the top-level format attribute should be placed on the whole struct on structs, and on variants on enums (we see an enum as an enumeration of some structs), while disallowed on a whole enum.

**Review by tyranron — CHANGES_REQUESTED** [review]
@ModProg at the moment, trait bounds for type parameters are being inferred incorrectly when top-level attribute comes into play. I've added tests covering those situations.

**Comment by JelteF** [comment]
> I've added tests covering those situations.

Somehow github is not showing them in this PR (but I see them on the branch on @ModProg their fork). I'm assuming the tests are not passing?

**Comment by ModProg** [comment]
@JelteF @tyranron I added it to the bounds_generation. If a format argument is present for the container, its bounds replace the field bounds.

**Review by tyranron — APPROVED** [review]
@ModProg thank you! Very nice job and quite a thoughtful implementation! 👍

I've ripped some repeated parts a little bit more, though, and made some bikeshedings regarding errors terminology.

@JelteF would you like to take a look, or we can merge it?

**Review by JelteF — APPROVED** [review]

**Comment by tyranron** [comment]
@JelteF could you make a `0.1.0-beta.3` release with fixes of recently merged PRs? I'd like to try it out on our codebase.

**Comment by JelteF** [comment]
> could you make a 0.1.0-beta.3 release with fixes of recently merged PRs?

Done

**Pull request merged**